### PR TITLE
Code for second lab

### PR DESCRIPTION
**Pull request by KhrystynaVysotska**

### CODE REVIEW

File: src/main/java/ua/lviv/iot/manager/Launcher.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+		int numberOfFoodPackagesAvailable = scanner.nextInt();
+		int numberOfHamstersInShop = scanner.nextInt();
+		Hamster[] hamsters = new Hamster[numberOfHamstersInShop];
+		for(int counter = 0; counter < numberOfHamstersInShop; counter++) {
+			if(scanner.hasNext()) {
+				hamsters[counter] = new Hamster(Integer.parseInt(scanner.next()), Integer.parseInt(scanner.next()), numberOfHamstersInShop - 1);
+			}
+		}
+		int maxMaxNumberOfHamsters = Seller.findMaxNumberOfHamsters(numberOfFoodPackagesAvailable, hamsters); 
+		String outputFileName = "D:\\AlgorithmLabs\\SecondLab\\hamsterAlgorithm\\hamstersOutput.txt";
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/manager/Launcher.java:25]: Старайтесь выносить такие вещи в конфигурацию или в аргументы запуска приложения (так же как и имя файла из которого вы читаете данные – т.о. эту программу можно будет запускать на других компах в том числе

File: src/main/java/ua/lviv/iot/manager/Launcher.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+		BufferedWriter writer = new BufferedWriter(new FileWriter(outputFileName));
+		writer.write(maxMaxNumberOfHamsters + "\ncounter: " + Seller.counter);
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/manager/Launcher.java:27]: Вывод данных в output файл не соответствует заданию. Если Вам нужно что-то дебажить – лучше писать на консоль или в error stream

File: src/main/java/ua/lviv/iot/model/Hamster.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package ua.lviv.iot.model;
+
+public class Hamster {
+
+	private int dailyNorm;
+	private int greedinessPerOneHamster;
+	private int totalFoodConsuming;
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/model/Hamster.java:7]: Когда Вы проектируете модели (классы), старайтесь абстрагироваться немного от задачи и задавать себе вопрос – вот эта штука нужна здесь?

В принципе, Хомяк может знать о своей полной прожорливости, но это не та информация которая должна в нём хранится. Например представьте, что одного хомяка перенесли из одной клетки в другую – его прожорливость изменилась, потому что кол-во соседей изменилось, но сам хомяк совсем не поменялся. В Вашем же случае при переносе хомяка нужно будет не забыть поменять ему значение полной прожорливости, иначе он останется с неверной информацией

File: src/main/java/ua/lviv/iot/model/Hamster.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	public Hamster(int dailyNorm, int greedinessPerOneHamster, int numberOfOtherHamsters) {
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/model/Hamster.java:8]: Тот же вопрос: должен ли хомяк знать сколько у него соседей при создании?

File: src/main/java/ua/lviv/iot/model/Hamster.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+		this.greedinessPerOneHamster = greedinessPerOneHamster;
+		setTotalFoodConsuming(numberOfOtherHamsters);
+	}
+	public int getDailyNorm() {
+		return dailyNorm;
+	}
+	public int getGreedinessPerOneHamster() {
+		return greedinessPerOneHamster;
+	}
+	public int getTotalFoodConsuming() {
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/model/Hamster.java:19]: А вот метод здесь кстати, но я бы его переформулировал. Как Вы думаете как?

File: src/main/java/ua/lviv/iot/seller/Seller.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
+package ua.lviv.iot.seller;
+
+import ua.lviv.iot.model.Hamster;
+
+public class Seller {
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/seller/Seller.java:5]: Зачем объявлять класс с private состоянием и делать всё в нём static? Static отличный концепт, когда он используется по назначению. Если Вы хотите хранить состояние – делайте обычный класс. Если Вам нужно хранить константы и набор утилитарных методов – ок, тогда можно static

File: src/main/java/ua/lviv/iot/seller/Seller.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	public static int counter = 0;
+	private static Hamster[] hamsters;
+	private static int numberOfFoodPackagesAvailable;
+
+	public static int findMaxNumberOfHamsters(int numberOfFoodPackagesAvailable, Hamster[] hamsters) {
+		Seller.hamsters = hamsters.clone();
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/seller/Seller.java:11]: Зачем?

File: src/main/java/ua/lviv/iot/seller/Seller.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+		Seller.numberOfFoodPackagesAvailable = numberOfFoodPackagesAvailable;
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/seller/Seller.java:12]: Аналогично. Зачем?

File: src/main/java/ua/lviv/iot/seller/Seller.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+		int length = hamsters.length;
+		int maxNumberOfHamsters = getMaxNumberOfHamsters(0, length - 1) + 1;
+		if(maxNumberOfHamsters == 0) {
+			counter++;
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/seller/Seller.java:16]: Кстати, а что это за counter? Для чего он нужен и что вообще считает?

File: src/main/java/ua/lviv/iot/seller/Seller.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	public static int counter = 0;
+	private static Hamster[] hamsters;
+	private static int numberOfFoodPackagesAvailable;
+
+	public static int findMaxNumberOfHamsters(int numberOfFoodPackagesAvailable, Hamster[] hamsters) {
+		Seller.hamsters = hamsters.clone();
+		Seller.numberOfFoodPackagesAvailable = numberOfFoodPackagesAvailable;
+		int length = hamsters.length;
+		int maxNumberOfHamsters = getMaxNumberOfHamsters(0, length - 1) + 1;
+		if(maxNumberOfHamsters == 0) {
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/seller/Seller.java:15]: Вот этой проверки я тоже не понял, честно говоря, т.е. Вы думаете, что Ваш алгоритм почему-то не работает с одним хомяком? Если так – это повод задуматься что пошло не так

File: src/main/java/ua/lviv/iot/seller/Seller.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+			counter++;
+			for(int i = 0; i < length; i++) {
+				if(hamsters[i].getDailyNorm() <= numberOfFoodPackagesAvailable) {
+					maxNumberOfHamsters++;
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/seller/Seller.java:19]: Я подозреваю, что Вы хотели сказать `maxNumberOfHamsters = 1`. ++ – операция относительно текущего значения, а Вы хотите записать абсолютное в переменную. И да сейчас они приводят к одному и тому же по случайности, но представьте, что Вы сейчас слегка поменяете код – вспомните ли Вы, что `maxNumberOfHamsters` должен быть 0?

File: src/main/java/ua/lviv/iot/seller/Seller.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+					break;
+				}
+			}
+		}
+		return maxNumberOfHamsters;
+	}
+	
+	private static int getMaxNumberOfHamsters(int startPivotPosition, int endPivotPosition) {
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/seller/Seller.java:27]: Просто совет: пока не научитесь читать алгоритмы быстро – комментируйте свой код. Желательно ДО написания кода проговорить свой алгоритм словами: что за чем должно выполняться. А затем комментировать код в соответствии с этим алгоритмом.

File: src/main/java/ua/lviv/iot/seller/Seller.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+		if(startPivotPosition > endPivotPosition) {
+			return startPivotPosition - 1;
+		}
+		int pivotPosition = findPivotPosition(startPivotPosition, endPivotPosition);
+		int numberOfFoodPackages = 0;
+		for (int hamsterPosition = 0; hamsterPosition <= pivotPosition; hamsterPosition++) {
+			counter++;
+			hamsters[hamsterPosition].setTotalFoodConsuming(pivotPosition);
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/seller/Seller.java:35]: Т.е. Вы здесь какой-то части хомяков по сути меняете количество соседей, а какой-то части оставляете то кол-во, что было раньше?

File: src/main/java/ua/lviv/iot/seller/Seller.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+		return leftPosition-1;
+	}
+
+	private static void swap(int currentPosition, int leftPosition) {
+		Hamster temp = hamsters[currentPosition];
+		hamsters[currentPosition] = hamsters[leftPosition];
+		hamsters[leftPosition] = temp;
+		
+	}
+}
```
Comment (PASSTor) [inline src/main/java/ua/lviv/iot/seller/Seller.java:66]: Я специально пока не комментирую сам алгоритм – во-первых Вам самой нужно до него дойти, во-вторых, я ещё не видел вторую итерацию вашего кода

File: src/test/java/ua/lviv/iot/seller/SellerTest.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
+package ua.lviv.iot.seller;
+
+import static org.junit.Assert.*;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import ua.lviv.iot.model.Hamster;
+
+public class SellerTest {
```
Comment (PASSTor) [inline src/test/java/ua/lviv/iot/seller/SellerTest.java:10]: Тесты. Это важно. Тесты должны тестировать не просто happy path, а и все возможные edge cases (экстремальные ситуации). Сейчас у Вас в тестах только те ситуации, которые были в примерах. Подумайте какие именно тесты нужны такому приложению, чтобы при любых изменениях кода убедиться, что алгоритм всё ещё верно работает.

File: src/test/java/ua/lviv/iot/seller/SellerTest.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+	private Hamster[] firstCaseHamsters = new Hamster[3];
+	private Hamster[] secondCaseHamsters = new Hamster[4];
+	private Hamster[] thirdCaseHamsters = new Hamster[2];
```
Comment (PASSTor) [inline src/test/java/ua/lviv/iot/seller/SellerTest.java:13]: ммм... А зачем сохранять что-то в приватную переменную класса, если это используется всего раз в одной функции?

File: src/test/java/ua/lviv/iot/seller/SellerTest.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+	@Before
+	public void initTestArray() {
+		firstCaseHamsters[0] = new Hamster(1, 2, 2);
+		firstCaseHamsters[1] =  new Hamster(2, 2, 2);
+		firstCaseHamsters[2] =  new Hamster(3, 1, 2);
+		
+		secondCaseHamsters[0] = new Hamster(5, 0, 3);
+		secondCaseHamsters[1] = new Hamster(2, 2, 3);
+		secondCaseHamsters[2] = new Hamster(1, 4, 3);
+		secondCaseHamsters[3] = new Hamster(5, 1, 3);
```
Comment (PASSTor) [inline src/test/java/ua/lviv/iot/seller/SellerTest.java:24]: Эти конструкции ужасны тем, что в них легко запутаться. Особенно при copy-paste. Постарайтесь использовать syntax sugar языка на котором пишете. Например посмотрите как объявляют массивы здесь: https://docs.oracle.com/javase/specs/jls/se7/html/jls-10.html#jls-10.2

**Review by PASSTor — APPROVED** [review]
